### PR TITLE
migration: unlock/unreserve Gov v1 balances and remove kvs

### PR DESCRIPTION
**Pull request by liamaharon**
Partial https://github.com/paritytech/polkadot/issues/6749

## Checklist

- [x] Restores Gov V1 pallet configurations removed in https://github.com/paritytech/polkadot/pull/6701
- [x] https://github.com/paritytech/substrate/pull/14226
- [x] https://github.com/paritytech/substrate/pull/14218
- [x] https://github.com/paritytech/substrate/pull/14241
- [x] Adds `RemovePallet` migration scaffolding, to be activated when pallets are removed from the runtime 
- [x] Disable calls to Gov V1 pallets by making them undecodable 
- [x] Wait for substrate migration PRs to merge
- [x] Activate unlock/unreserves for Polkadot

## Migration summary

| Network | Module | Total Accounts | Total Stake to Unlock | Total Deposit to Unreserve | USD Total |
|---------|--------|----------------|----------------------|----------------------------|----------------|
| Kusama  | Democracy | 5364 | 1,179,551 KSM | 25 KSM | $30,609,997 |
| Kusama  | ElectionPhragmen | 748 | 1,496,706 KSM | 157 KSM | $38,843,594 |
| Kusama  | Tip | 3 | N/A | 0.21 KSM | $5 |
| Polkadot | Democracy | 5228 | 44,984,449 DOT | 800 DOT | $260,914,444 |
| Polkadot | ElectionPhragmen | 326 | 235,329,403 DOT | 5,649 DOT | $1,364,943,301 |
| Polkadot | Tip | 7 | N/A | 16 DOT | $92 |

assuming DOT: $5.8, KSM: $25.95

## Questions / to figure out

- [x] Total to unstake amounts seem quite high, but not necessarily invalid. Are there any dashboards we can cross-check them against?~
- [x] What's the best way to disable calls to Gov V1 pallets, incl. if they're made as part of batch requests?
- [x] `ERROR main runtime::system: Logic error: Unexpected underflow in reducing consumer` error logs
    - See [this comment](https://github.com/paritytech/polkadot/pull/7314#issuecomment-1589933081)

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2920843

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2921510

File: runtime/kusama/src/lib.rs [hunk 1]
```diff
@@ -1499,6 +1518,16 @@ impl Get<Perbill> for NominationPoolsMigrationV4OldPallet {
 	}
 }
 
+// Gov V1 pallets pending storage removal using the frame_support RemovePallet migration.
+parameter_types! {
+	pub const DemocracyStr: &'static str = "Democracy";
```
Comment (kianenigma) [inline runtime/kusama/src/lib.rs:1523]: wondering when static str can be used as const-generics, which would allow us to have one `ConstStr<const S: &'static str>` or something like that.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2980189

**Comment by ggwpez** [comment]
What is this in the Kusama try-runtime log?  
```pre
ERROR main runtime::system: Logic error: Unexpected underflow in reducing consumer

**Comment by liamaharon** [comment]
> What is this in the Kusama try-runtime log? Looks like a few hundreds of those.
> 
> ```
> ERROR main runtime::system: Logic error: Unexpected underflow in reducing consumer
> ```

It's benign. 

@gpestana suggested [in this comment](https://github.com/paritytech/substrate/pull/14218#discussion_r1209583542) that it's likely to be related to https://github.com/paritytech/substrate/issues/13114#issuecomment-1377697085. 

I was unable to log the offending accounts to check their state exactly (see the comment), but I did confirm that all the accounts "Unexpected underflow in reducing consumer" occurs for have real locks in Balances, so it's necessary to `T::Currency::remove_lock` on those accounts even if it triggers that log, otherwise it'll leave hanging locks in the Balances.

File: runtime/kusama/src/lib.rs [hunk 2]
```diff
@@ -1559,7 +1605,32 @@ pub mod migrations {
...
 			}
 
-			RocksDbWeight::get().reads_writes(2, 2)
+			let storage_version = Democracy::on_chain_storage_version();
+			if storage_version < 1 {
+				StorageVersion::new(1).put::<Democracy>();
+			}
+
+			let storage_version = Council::on_chain_storage_version();
+			if storage_version < 4 {
```
Comment (ggwpez) [inline runtime/kusama/src/lib.rs:1614]: Why is this needed?  
I thought the pallet storage is still there.

File: runtime/kusama/src/lib.rs [hunk 3]
```diff
@@ -1533,6 +1562,23 @@ pub mod migrations {
...
+		// Gov v1 storage migrations
+		// https://github.com/paritytech/polkadot/issues/6749
+		pallet_elections_phragmen::migrations::unlock_and_unreserve_all_funds::UnlockAndUnreserveAllFunds<Runtime>,
+		pallet_democracy::migrations::unlock_and_unreserve_all_funds::UnlockAndUnreserveAllFunds<Runtime>,
+		pallet_tips::migrations::unreserve_deposits::UnreserveDeposits<Runtime, ()>,
+
+		// RemovePallets only after they have been removed from the runtime. Otherwise, the on-chain
+		// storage version is removed for active pallets causing try-runtime to fail. The below code
+		// should be uncommented when the pallets are removed from the runtime.
+		// frame_support::migrations::RemovePallet<DemocracyStr, RocksDbWeight>,
```
[the remark this answers is not in the document]
Comment (liamaharon) [reply runtime/kusama/src/lib.rs:1575]: Yes, next upgrade when we remove the pallets from the runtime. 

This will be either 1.1 or 1.2 (depending if this makes it into 1.0), I'll update the comment with more info.

File: runtime/kusama/src/lib.rs [hunk 4]
```diff
@@ -1533,6 +1562,23 @@ pub mod migrations {
 	);
 	pub type V0943 = (
 		SetStorageVersions,
+
```
[the remark this answers is not in the document]
Comment (liamaharon) [reply runtime/kusama/src/lib.rs:1565]: Yeah, looks like the auto resolution of merge conflicts put all my stuff in .43. Will fix.

File: runtime/kusama/src/lib.rs (hunk 2 above)
Comment (liamaharon) [reply runtime/kusama/src/lib.rs:1614]: I swore that the on-chain versions had not been set causing the on-runtime-upgrade to fail when I added these, but I can't reproduce it now.. Probably a mistake, I'll remove them.

**Comment by liamaharon** [comment]
bot fmt

**Comment by liamaharon** [comment]
bot rebase

**Comment by ggwpez** [comment]
This should fix the polkadot try-runtime test. So we can check it again before releasing https://polkadot.polkassembly.io/referenda/9

**Comment by paradox-tt** [comment]
Hey devs,

I've got two questions:

1. I have some DOT that's caught up with [this bug](https://substrate.stackexchange.com/questions/5067/delegating-and-undelegating-during-the-lock-period-extends-it-for-the-initial-am/5232#5232).  Would all DOT locked due to delegation also be freed when this is implemented on Polkadot?
2. Would any KSM/DOT reserved due to Treasury Spend Proposals also be freed?

**Comment by liamaharon** [comment]
> Hey devs,
> 
> I've got two questions:
> 
> 1. I have some DOT that's caught up with [this bug](https://substrate.stackexchange.com/questions/5067/delegating-and-undelegating-during-the-lock-period-extends-it-for-the-initial-am/5232#5232).  Would all DOT locked due to delegation also be freed when this is implemented on Polkadot?
> 2. Would any KSM/DOT reserved due to Treasury Spend Proposals also be freed?  I am asking this specifically as there are some proposers seeking refund of their deposits via OpenGov.

Hey @paradox-tt,

1. Yes, all funds from the Democracy pallet will be unlocked. If you let me know an address with the bugged funds I could confirm for you if you like.

2. Yeah I believe they will be.. Could you please link to the OpenGov vote?

**Comment by paradox-tt** [comment]
> > Hey devs,
> > I've got two questions:
> > 
> > 1. I have some DOT that's caught up with [this bug](https://substrate.stackexchange.com/questions/5067/delegating-and-undelegating-during-the-lock-period-extends-it-for-the-initial-am/5232#5232).  Would all DOT locked due to delegation also be freed when this is implemented on Polkadot?
> > 2. Would any KSM/DOT reserved due to Treasury Spend Proposals also be freed?  I am asking this specifically as there are some proposers seeking refund of their deposits via OpenGov.
> 
> Hey @paradox-tt,
> 
> 1. Yes, all funds from the Democracy pallet will be unlocked. If you let me know an address with the bugged funds I could confirm for you if you like.
> 2. Yeah I believe they will be.. Could you please link to the OpenGov vote?

My apologies for my tardy reply,

1. The account with the bugged funds is `14hM4oLJCK6wtS7gNfwTDhthRjy5QJ1t3NAcoPjEepo9AH67`
2. Here's [an example](https://polkadot.subsquare.io/referenda/referendum/25) of someone requesting a refund for deposit funds locked with [Treasury Spend Proposal 282](https://polkadot.subsquare.io/treasury/proposal/282)

With regards to the second item, mention was made in the OP that funds would be released from tips, democracy and council votes but no mention was made wrt deposits within the treasury pallet.

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/psa-gov-v1-funds-will-be-automatically-refunded-next-runtime-upgrade/3270/1

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3103198

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3103265

**Comment by thekiwimango** [comment]
Hey devs, any updates here? Our KSM has been locked for quite a while now with no option to unlock it

**Comment by liamaharon** [comment]
> Hey devs, any updates here? Our KSM has been locked for quite a while now with no option to unlock it

It's pending review, hoping to include in the next runtime upgrade.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3243259

File: runtime/kusama/Cargo.toml [hunk 5]
```diff
@@ -71,7 +71,7 @@ pallet-referenda = { git = "https://github.com/paritytech/substrate", branch = "
 pallet-scheduler = { git = "https://github.com/paritytech/substrate", branch = "master", default-features = false }
 pallet-session = { git = "https://github.com/paritytech/substrate", branch = "master", default-features = false }
 pallet-society = { git = "https://github.com/paritytech/substrate", branch = "master", default-features = false, features = ["experimental"] }
-frame-support = { git = "https://github.com/paritytech/substrate", branch = "master", default-features = false }
+frame-support = { git = "https://github.com/paritytech/substrate", branch = "master", default-features = false, features = ["tuples-96"] }
```
Comment (liamaharon) [inline runtime/kusama/Cargo.toml:74]: Compilation fails without this. Compile error tells me to add this feature.

File: runtime/kusama/src/lib.rs [hunk 6]
```diff
@@ -1390,6 +1404,15 @@ construct_runtime! {
 		ImOnline: pallet_im_online::{Pallet, Call, Storage, Event<T>, ValidateUnsigned, Config<T>} = 11,
 		AuthorityDiscovery: pallet_authority_discovery::{Pallet, Config<T>} = 12,
 
+		// Old, deprecated governance pallets. Here purely so storage can be migrated.
+		// See https://github.com/paritytech/polkadot/issues/6749
+		Democracy: pallet_democracy exclude_parts {Call} = 13,
```
Comment (bkchr) [inline runtime/kusama/src/lib.rs:1409]: I don't think that we need to add the pallets to construct runtime? Implementing the config traits and then referencing the migrations should be enough?

Comment (liamaharon) [reply runtime/kusama/src/lib.rs:1409]: This was discussed in Element some time ago, removing them leads to compilation issues. 

It was suggested instead to leave them in the runtime, but exclude_parts the `Call` to ensure the pallets are inoperable. 

Screenshot 2023-05-31 at 13 36 18

**Review by ggwpez — APPROVED** [review]

**Review by kianenigma — APPROVED** [review]

File: runtime/kusama/src/governance/old.rs [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+	pub const MinimumDeposit: Balance = 100 * CENTS;
+	pub EnactmentPeriod: BlockNumber = prod_or_fast!(8 * DAYS, 1, "KSM_ENACTMENT_PERIOD");
+	pub CooloffPeriod: BlockNumber = prod_or_fast!(7 * DAYS, 1 * MINUTES, "KSM_COOLOFF_PERIOD");
+	pub const InstantAllowed: bool = true;
+	pub const MaxVotes: u32 = 100;
+	pub const MaxProposals: u32 = 100;
+	pub MaxProposalWeight: Weight = Perbill::from_percent(50) * BlockWeights::get().max_block;
+}
+
+impl pallet_democracy::Config for Runtime {
```
Comment (rphmeier) [inline runtime/kusama/src/governance/old.rs:44]: shouldn't the "old" democracy pallet be as close to unactivated as possible?

i.e. all the origins should be `EnsureNever`.

Comment (liamaharon) [reply runtime/kusama/src/governance/old.rs:44]: Hmm I had not considered the hooks. You're right, ideally they would not run.

The issue wasn't with accessing storage but something a bit nastier, I'll revisit tomorrow.

Comment (liamaharon) [reply runtime/kusama/src/governance/old.rs:44]: I'm not sure what you mean "implementing the configs for a made up `struct DepRuntime`", could you please elaborate?

File: runtime/kusama/src/lib.rs (hunk 6 above)
Comment (gupnik) [reply runtime/kusama/src/lib.rs:1409]: I believe the way to do it properly would be to create a dummy RuntimeEvent that implements the `From` and pass that instead of the actual RuntimeEvent.

Something like
```rust
struct DepRuntimeEvent;
impl From<pallet_democracy::Event<Runtime>> for DepRuntimeEvent {
	fn from(_: pallet_democracy::Event<Runtime>) -> Self {
		Self
	}
}
```

This can be entirely dummy, just enough to make the trait bounds pass I think. 

You would probably need a similar RuntimeOrigin and RuntimeCall.

**Comment by liamaharon** [comment]
@rphmeier @bkchr this PR was discussed in our team FRAME call yesterday.

It's non-trivial to remove the pallets from `construct_runtime!`. We decided to not spend too much bandwidth for such a one-time fix, especially since merging this is time sensitive. 

Instead we agreed to temporarily restore the pallets in the runtime, and disable `Call`s to the pallets and the hooks from running.

- `Call`s have been disabled already with `exclude_parts`. 
- We agreed to add a feature flag which would allow disabling the hooks we don't want to run. There may be better approaches to disabling hooks (such as creating a new part we can use in exclude_parts), but that would likely take a long time to implement and unlocking the Gov V1 funds is time sensitive.
    - PR adding a new feature flag to pallets with consequential hooks: https://github.com/paritytech/substrate/pull/14708

File: runtime/kusama/src/lib.rs (hunk 6 above)
Comment (liamaharon) [reply runtime/kusama/src/lib.rs:1409]: This unfortunately didn't work, there're more errors. Welcome to try it yourself if you like. Otherwise we can pursue the `disable-hooks` approach.

Comment (bkchr) [reply runtime/kusama/src/lib.rs:1409]: > This was discussed in Element some time ago, removing them leads to compilation issues. 
> 
> It was suggested instead to leave them in the runtime, but `exclude_parts` the `Call` to ensure the pallets are inoperable. 
> 
> Screenshot 2023-05-31 at 13 36 18
> 

You can just implement the requested trait bound yourself.

Comment (liamaharon) [reply runtime/kusama/src/lib.rs:1409]: This is not so trivial and led to more compiler issues, but I'll have another go and ask for help if I get stuck again.

**Comment by liamaharon** [comment]
Basti to the rescue: https://github.com/paritytech/substrate/pull/14773#pullrequestreview-1579975487

I am going to open a Substrate PR modifying the other two migrations to be completely Runtime independent. Once that is merged, I can finally update this PR to not require the pallets in the runtime and there should no longer be any blockers to merging this.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3395996

File: runtime/polkadot/src/lib.rs [hunk 8]
```diff
@@ -1376,9 +1376,9 @@ construct_runtime! {
 		TechnicalCommittee: pallet_collective::<Instance2>::{Pallet, Call, Storage, Origin<T>, Event<T>, Config<T>} = 16,
 		PhragmenElection: pallet_elections_phragmen::{Pallet, Call, Storage, Event<T>, Config<T>} = 17,
 		TechnicalMembership: pallet_membership::<Instance1>::{Pallet, Call, Storage, Event<T>, Config<T>} = 18,
-		Treasury: pallet_treasury::{Pallet, Call, Storage, Config<T>, Event<T>} = 19,
 
 		// OpenGov stuff.
+		Treasury: pallet_treasury::{Pallet, Call, Storage, Config<T>, Event<T>} = 19,
```
Comment (liamaharon) [inline runtime/polkadot/src/lib.rs:1381]: Can someone confirm `Treasury` needs to stay in OpenGov?

**Comment by liamaharon** [comment]
Needs https://github.com/paritytech/substrate/pull/14779. Then, should be no more blockers 🤞

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3400552

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3404917

**Review by bkchr — APPROVED** [review]
Ty! Much better approach! :)

File: runtime/polkadot/src/lib.rs [hunk 9]
```diff
@@ -202,11 +201,6 @@ impl PrivilegeCmp<OriginCaller> for OriginPrivilegeCmp {
 		match (left, right) {
 			// Root is greater than anything.
 			(OriginCaller::system(frame_system::RawOrigin::Root), _) => Some(Ordering::Greater),
-			// Check which one has more yes votes.
-			(
-				OriginCaller::Council(pallet_collective::RawOrigin::Members(l_yes_votes, l_count)),
-				OriginCaller::Council(pallet_collective::RawOrigin::Members(r_yes_votes, r_count)),
-			) => Some((l_yes_votes * r_count).cmp(&(r_yes_votes * l_count))),
```
Comment (liamaharon) [inline runtime/polkadot/src/lib.rs:209]: Copying Gav from here https://github.com/paritytech/polkadot/pull/6701/files#diff-db5840c3c03c740de25aed56b9257559c224a04619c079c94cddbb23f9a855a4L210-L214

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/3410819

**Review by bkchr — APPROVED** [review]

**Review by ggwpez — APPROVED** [review]

**Comment by liamaharon** [comment]
bot merge force
